contracts/contracts/strategies/aerodrome/AerodromeAMOStrategy.sol

```diff
@@ -0,0 +1,964 @@
+     *             tick ranges to the pool to facilitate the swap. Effectively turning Scenario 1 into a Scenario 2
+     * Scenario 2: When there is no liquidity in the pool from the strategy but there is from other LPs then
+     *             only step 1 is skipped. (It is important to note that liquidity needs to exist in the configured
+     *             strategy tick ranges in order for the swap to be possible) Step 3 mints new liquidity position
+     *             instead of adding to an existing one.
+     * Scenario 3: When there is strategy's liquidity in the pool all 3 steps are taken
+     *
+     *
+     * Exact _amountToSwap, _minTokenReceived & _swapWeth parameters shall be determined by simulating the
+     * transaction off-chain. The strategy checks that after the swap the share of the tokens is in the
+     * expected ranges.
+     *
+     * @param _amountToSwap The amount of the token to swap
+     * @param _swapWeth Swap using WETH when true, use OETHb when false
+     * @param _minTokenReceived Slippage check -> minimum amount of token expected in return
+     */
+    function rebalance(
+        uint256 _amountToSwap,
+        bool _swapWeth,
+        uint256 _minTokenReceived
+    ) external nonReentrant onlyGovernorOrStrategist {
+        /**
+         * Would be nice to check if there is any total liquidity in the pool before performing this swap
+         * but there is no easy way to do that in UniswapV3:
+         * - clPool.liquidity() -> only liquidity in the active tick
+         * - asset[1&2].balanceOf(address(clPool)) -> will include uncollected tokens of LP providers
+         *   after their liquidity position has been decreased
+         */
+
+        /**
+         * When rebalance is called for the first time there is no strategy's
+         * liquidity in the pool yet. The full liquidity removal is thus skipped.
+         */
+        if (tokenId != 0) {
+            _removeLiquidity(1e18);
+        }
+        // in some cases we will just want to add liquidity and not issue a swap to move the
+        // active trading position within the pool
+        if (_amountToSwap > 0) {
+            _swapToDesiredPosition(_amountToSwap, _minTokenReceived, _swapWeth);
+        }
+        // calling check liquidity early so we don't get unexpected errors when adding liquidity
+        // in the later stages of this function
+        _checkForExpectedPoolPrice();
+
+        _addLiquidity();
+        _checkForExpectedPoolPrice();
+    }
+
+    // slither-disable-end reentrancy-no-eth
+
+    /**
+     * @dev Decrease partial or all liquidity from the pool.
+     * @param _liquidityToDecrease The amount of liquidity to remove expressed in 18 decimal point
+     */
+    function _removeLiquidity(uint256 _liquidityToDecrease)
+        internal
+        gaugeUnstakeAndRestake
+    {
+        require(_liquidityToDecrease > 0, "Must remove some liquidity");
```

Comment:
Rather than `_liquidityToDecrease > 0`, would it be better to check the computed `liquidityToRemove > 0` (along with `liquidity > 0`)

Reply:
Agreed added the check: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/4dce8886b2e90d3d204b6e51942d8db4db10e720

Reply:
Actually this was a bad idea and it introduced an error. There are valid cases where `liquidity` == 0 and consequently `liquidityToDecrease` is also 0: 
- after withdrawAll is called
- aftrer withdraw removes all of the liquidity

Depositing tokens and rebalancing after the withdrawal would thus fail. Here is the fix and a fork test: https://github.com/OriginProtocol/origin-dollar/pull/2184/commits/0b31bcdc7f75bb112ec884b619fde1b39dfd4877

Reply:
cc @shahthepro @DanielVF ☝️